### PR TITLE
Address / is valid

### PR DESCRIPTION
**Pull request by gimre-xymcity**

### CODE REVIEW

File: symbolchain/core/Network.py [hunk 1]
```diff
@@ -28,6 +28,21 @@ def public_key_to_address(self, public_key):
...
+        if address.bytes[0] != self.identifier:
+            return False
+
+        hash_builder = self.address_hasher()
+        hash_builder.update(address.bytes[0:1 + 20])
+
+        checksum = address.bytes[1+20:]
+        actual_checksum = hash_builder.digest()[0:len(checksum)]
+        if checksum != actual_checksum:
+            print('mismatch {} vs {}'.format(checksum, actual_checksum))
```
Comment (gimre-xymcity) [inline symbolchain/core/Network.py:42]: remove I guess?

Comment (Jaguar0625) [reply symbolchain/core/Network.py:42]: yes, please remove prints :)

File: symbolchain/core/Network.py [hunk 2]
```diff
@@ -28,6 +28,21 @@ def public_key_to_address(self, public_key):
...
+    def is_valid_address(self, address):
+        """Validates the address."""
+        if address.bytes[0] != self.identifier:
+            return False
+
+        hash_builder = self.address_hasher()
+        hash_builder.update(address.bytes[0:1 + 20])
+
+        checksum = address.bytes[1+20:]
+        actual_checksum = hash_builder.digest()[0:len(checksum)]
```
Comment (Jaguar0625) [inline symbolchain/core/Network.py:40]: not sure but i might go with more verbose names, e.g. checksum_from_address,  calculated_checksum

File: tests/test/BasicNetworkTest.py [hunk 3]
```diff
@@ -16,6 +16,21 @@ def __init__(
         (self.expected_testnet_address, self.testnet_network) = expected_testnet_address_network_pair
 
 
+def mainnet_accessor(descriptor):
+    return descriptor.mainnet_network
+
+
+def testnet_accessor(descriptor):
+    return descriptor.testnet_network
```
Comment (Jaguar0625) [inline tests/test/BasicNetworkTest.py:24]: don't mind, but you technically don't need these; could just pass name string and use getattr

File: tests/test/BasicNetworkTest.py [hunk 4]
```diff
@@ -16,6 +16,21 @@ def __init__(
...
+
+
+def split_at(buffer, position):
+    if position < 0:
+        position = len(buffer) + position
+    assert 0 <= position < len(buffer)
+    return [buffer[0:position], buffer[position:position+1], buffer[position+1:]]
```
Comment (Jaguar0625) [inline tests/test/BasicNetworkTest.py:31]: any particular reason you didn't want this to do mutate as well?

File: tests/test/BasicNetworkTest.py [hunk 5]
```diff
@@ -39,6 +54,44 @@ def test_can_convert_testnet_public_key_to_address(self):
...
+
+        # Act + Assert:
+        self.assertTrue(network_accessor(test_descriptor).is_valid_address(address))
+
+    def _test_validate_invalid_address(self, network_accessor, position):
+        # Arrange:
+        test_descriptor = self.get_test_descriptor()
+        address = network_accessor(test_descriptor).public_key_to_address(test_descriptor.deterministic_public_key)
+
+        # mutate last byte
```
Comment (Jaguar0625) [inline tests/test/BasicNetworkTest.py:88]: add `- `

File: tests/test/BasicNetworkTest.py [hunk 6]
```diff
@@ -39,6 +54,44 @@ def test_can_convert_testnet_public_key_to_address(self):
...
+        parts = split_at(address.bytes, position)
+        address.bytes = parts[0] + (parts[1][0] ^ 0xFF).to_bytes(1, 'little') + parts[2]
```
Comment (Jaguar0625) [inline tests/test/BasicNetworkTest.py:90]: `(parts[1][0] ^ 0xFF).to_bytes(1, 'little')` => `[parts[1][0] ^ 0xFF]` ?

**Review by Jaguar0625 — CHANGES_REQUESTED** [review]

File: tests/test/BasicNetworkTest.py (hunk 6 above)
Comment (gimre-xymcity) [reply tests/test/BasicNetworkTest.py:90]: _can't concat list to bytes_

File: symbolchain/core/Network.py (hunk 1 above)
Comment (gimre-xymcity) [reply symbolchain/core/Network.py:42]: I wasn't sure if it should raise exception or return error, but I'm guessing error is better

Comment (Jaguar0625) [reply symbolchain/core/Network.py]: it would just return false, wouldn't it?

**Review by Jaguar0625 — APPROVED** [review]

**Pull request merged**